Add clang-tidy CI step

ZehMatt opened the pull request:
I've tried a few actions from the GitHub Marketplace and ended up writing my own custom action, it can be found here https://github.com/ZehMatt/clang-tidy-test

What this PR does: It adds a new workflow called clang-tidy-test which will run only on pull requests as it needs the information from the pull such as which files are modified and which lines. The custom action runs clang-tidy only on the modified files and also only shows errors/warnings on the modified lines. I enabled only a few clang-tidy checks for now as seen in the .clang-tidy file. The action is set to fail if there are any clang-tidy issues and will annotate the lines with the error/warning message and must be resolved by the user. 

I've added a test commit which I will remove if we plan to merge this.

tupaschoal commented:
I think I know where you want to get, but I'll mark this as draft until you post some reasoning, I hope that's ok!

ZehMatt commented:
> I think I know where you want to get, but I'll mark this as draft until you post some reasoning, I hope that's ok!

Well one of the reasons is to make reviews a bit easier, one of the checks looks for macros used as constants which we definitely don't want and its best if a computer complains rather than a person. The other thing is that this can potentially spot issues that are easily missed by any reviewer, but it makes sense to be in draft until it all works.

ZehMatt commented:
Alright, its now working properly, you can see it annotate the files by navigating to files or simply check the summary of the failed action.

duncanspumpkin commented on .github/workflows/clang-tidy.yml:
```diff
@@ -0,0 +1,38 @@
+name: clang-tidy-test
```
Clang tidy check

tupaschoal commented on .github/workflows/clang-tidy.yml:
```diff
@@ -0,0 +1,34 @@
+    runs-on: ubuntu-latest
+    container: openrct2/openrct2-build:12-jammy
+    # This job requires a pull request event to run.
+    if: github.event_name == 'pull_request'
```
Is this needed, since the job already applies only to on pull requests?

ZehMatt replied:
No, ill remove it, was initially in the CI workflow.

tupaschoal commented on .github/workflows/clang-tidy.yml:
```diff
@@ -0,0 +1,34 @@
+    - uses: ZehMatt/clang-tidy-annotations@v1.0.0
+      with:
+        build_dir: 'build'
+        cmake_args: '-G Ninja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_BUILD_TYPE=Debug -DDISABLE_DISCORD_RPC=ON -DDISABLE_NETWORK=ON -DDISABLE_HTTP=ON -DDISABLE_FLAC=ON -DDISABLE_VORBIS=ON -DDISABLE_OPENGL=ON'
```
Should we disable all this? Won't it make some of the code "uncovered"?

ZehMatt replied:
Yeah, I was initially testing this on the barebones ubuntu image but our own image has all the deps, I will however keep discord disabled, that isn't covered in the container image.

janisozaur commented on .github/workflows/clang-tidy.yml:
```diff
@@ -0,0 +1,32 @@
+    - name: Get clang-tidy
+      run: |
+        apt-get update
+        apt-get install -y clang-tidy
```
I will update our base image to include that, probably later this week

ZehMatt replied:
Great, thanks.

ZehMatt commented:
> Looks ok to me, I'm just concerned if it will warn on non-changed code and have people fix things unrelated to what they are changing

It will filter as much as possible to only lines affected, its possible it triggers something on lines close by but that's why I also only enabled a very limited set of checks at the moment.

tupaschoal commented:
Ok

Gymnasiast commented on .clang-tidy:
```diff
@@ -0,0 +1,9 @@
+    cppcoreguidelines-macro-usage,
+    modernize-use-nullptr,
+    modernize-use-override,
+    readability-avoid-const-params-in-decls
```
If I understand it correctly, this will disallow adding const when parameters are passed by value anyway,  like `const uint8_t spam`. Is that correct?

ZehMatt replied:
Also that reminded me to have https://clang.llvm.org/extra/clang-tidy/checks/readability/const-return-type.html added in the next batch of checks.

duncanspumpkin replied:
Can we have the opposite. It's better to have the const in the declaration then you know if a parameter is modified. The definition can drop the const if they want since it isn't part of the signature.

ZehMatt replied:
The parameter can not be modified unless its a reference, for references it won't say anything.

Gymnasiast replied:
@ZehMatt That is not quite what I asked... does it _only_ affect cases where the parameter is passed by value? I already found the page you linked to myself, but didn’t find it clear.

ZehMatt replied:
Here is a more concrete example to when it applies:
```cpp
// Complains because const is meaningless here
void func(const int arg1);

// Won't complain.
void func(const int& arg1);

// Won't complain, here const makes sense if you don't want it to mutate within the function body, still matches the signature of void func(const int arg1);
void func(const int arg1) { ... }
```

duncanspumpkin replied:
Yes so I'd like it to be the opposite. It should complain on `void func(int arg1);` to improve readability you want const to be on all interfaces. Then a non const reference is very obvious. The implementation can drop the const if you want but don't care as much.

ZehMatt replied:
But having const in the declaration does nothing really, when its passed by value then there are no side effects, so why do you want const there?

Gymnasiast replied:
> Here is a more concrete example to when it applies:
> 
> ```c++
> // Complains because const is meaningless here
> void func(const int arg1);
> 
> // Won't complain.
> void func(const int& arg1);
> 
> // Won't complain, here const makes sense if you don't want it to mutate within the function body, still matches the signature of void func(int arg1);
> void func(const int arg1) { ... }
> ```

So does it leave classes and structs as parameters alone?

ZehMatt replied:
When its a const reference yes.

duncanspumpkin replied:
I'd rather we don't enable this particular rule.

ZehMatt replied:
I'll remove it.

ZehMatt commented:
I removed the test commit and removed the one check for const args in declarations, its only checking now for macros used as constants. We can look into adding more useful things later on and also enforce our coding style with it but that can be discussed internally on Discord.
